### ISSUE TITLE
[Bug] Contract Trapped Issue due to Ordering of Struct

### ISSUE THREAD

**Issue by shirshak55** [paritytech/ink#568]
Currently in https://gist.github.com/shirshak55/504745cc62539ae5f6388491448ae19b if you try to invoke get_entry two times you will get Contract Trapped.

So if you follow println you will see following image giving some false indication like `vector.len()` cannot be called 
https://pasteboard.co/JyPBED9.png

But the error was due to ordering of field which I thought wouldn't matter in Rust and its normal in Rust code.

But if we use this struct it doesn't give Contract Trapped Error.

```rust
#[ink(storage)]
pub struct Charity {
    // User can play only once and added to pool with one submission. HashMap is perfect fit here.
    // users_vec so that we can choose it easily when we draw it
    users: HashMap<User, Balance>,
    users_vec: Vec<User>,
    // The collected money is sent  to predefined address. Which is the one who calls the contract first time
    collector: AccountId,
    // Total amount collected till now so we don't have to iterate over users to know it.
    // Just for optimization
    amount_collected: Balance,
    // Winners List
    winners: Vec<User>,
    countdown: Option<Timestamp>,
}
```

I would like to thank Robin from Discord for figuring out ordering issue. I had already wasted like 24 hour and if he weren't there i bet I wouldn't have found bug for another 24 hour :) Thanks to him.

According to him:

> the `Option<Timestamp>` is suspicious because we do some arcane optimized handling for Option<T> values in ink_storage to avoid some storage reads and writes and apparently things that are complex tend to have bugs.

**Comment by Robbepop** [comment]
Thanks for the bug report! We are working on a fix. :)

**Comment by cmichi** [comment]
Thanks for the report!

This is the minimal contract to reproduce the bug:
```rust
#![cfg_attr(not(feature = "std"), no_std)]

use ink_lang as ink;

#[ink::contract]
mod bug {
    use ink_storage::collections::Vec;

    #[ink(storage)]
    pub struct Bug {
        countdown: Option<u32>,
        winners: Vec<u32>,
    }

    impl Bug {
        #[ink(constructor)]
        pub fn new() -> Self {
            Self {
                countdown: None,
                winners: Vec::new(),
            }
        }

        #[ink(message)]
        pub fn bug(&mut self) -> Result<(), ()> {
            ink_env::debug_println("get len");

            // on the second call of `bug()` we panic here at the `len()` lookup.
            // the error is `ExtError::KeyNotFound`.
            if self.winners.len() == 2 {
                ink_env::debug_println("gotten len");
                return Err(());
            }

            ink_env::debug_println("is_none?");
            if self.countdown.is_none() {
                ink_env::debug_println("setting countdown");
                self.countdown = Some(13);
            }

            ink_env::debug_println("end");
            Ok(())
        }
    }
}
```

**Comment by shirshak55** [comment]
Thanks :)

**Comment by Robbepop** [comment]
Will be included in the next ink 3.0-rc3 release.
